nova/compute/manager.py

```diff
@@ -10680,6 +10681,35 @@ def _reclaim_queued_deletes(self, context):
                                 "instance: %s",
                                 e, instance=instance)
 
+    @periodic_task.periodic_task(
+            spacing=CONF.instances_stuck_in_deleting_cleanup_interval)
+    def _cleanup_instances_stuck_in_deleting(self, context):
+        """Reclaim instances stuck in DELETING state, but not really destroyed,
+        and retrigger terminate_instance for them.
+        """
+        filters = {'task_state': task_states.DELETING,
+                   'host': self.host}
+        instances = objects.InstanceList.get_by_filters(
+            context, filters,
+            expected_attrs=objects.instance.INSTANCE_DEFAULT_FIELDS,
+            use_slave=True)
+        for instance in instances:
+            # Check if the instance has a process lock
+            for name, s in _semaphores._semaphores.items():
+                if name == instance.uuid:
+                    break
+            else:
+                LOG.info(
+                        "Cleanup instance found in DELETING state but "
+                        "has no process lock", instance=instance)
+                try:
+                    bdms = objects.BlockDeviceMappingList.get_by_instance_uuid(
+                            context, instance.uuid)
+                    self.terminate_instance(context, instance, bdms)
```

Comment:
Is it possible that somebody changes the `BlockDeviceMappingList` while we wait for the lock in `terminate_instance()`? What happens then?

Reply:
If i got your question correctly, you mean if another function locks the instance and `terminate_instance` waits for this lock to be released and meanwhile something changed in the `bdms`, yes this condition is not handled exactly in the original `terminate_instance` function but there is a similar approach for that, if you think that is something likely to happen we can add a parameter to `terminate_instance` to enforce it to refresh the `BlockDeviceMappingList` if needed, and we pass this parameter to call in our task
does that sound reasonable ?